Cosmic flux fix

arturof opened the pull request:
Cosmic muons were always crossing exactly through the centre of the detector. Added code so they cross a random point uniformly distributed in the detector volume.

Fixed bug where cosmic histograms content was shifted by 1 bin and another one where the GetBinContent function was incorrectly used to find the muon mean energy.

Run was crashing because cosmic muons were generated outside the world volume and when deleting the cosmic histograms in the destructor. Both issues are fixed now.

arturof committed:
Fixed bin numbers when setting or getting content
of cosmic muon histograms

arturof committed:
Generate cosmic muon at a fixed distance from the
center of the detector and passing through a point
uniformly distributed in the ID (instead of the center)

arturof committed:
Merge branch 'WCSim:develop' into cosmic-flux-fix

arturof committed:
Reduced altitude variable so cosmic muons start inside the world
volume (experimental hall) but still outside the detector.
Otherwise run crashes.

arturof committed:
Detach cosmics histograms from flat ntuple file to avoid crash
when deleting histogram pointers in the destructor

tdealtry commented:
Looks great. Thanks @arturof 
Since the CI doesn't check cosmics automatically, I wonder if you can post a plot (or few) to show this is working as expected. I presume you've made some already for a talk?

arturof commented:
Yes, I have a few plots.

WCSim converts the input cosmic flux text file (data/MuonFlux-HyperK-ThetaPhi.dat by default) in histograms that are saved in a file named cosmicflux.root. The histograms are the flux and mean energy as a function theta and phi angles. The last bin of the histograms were empty because the first bin in the text file is 0 instead of 1 as in ROOT convention. I fixed that and they look like this now:

![hFluxCosmics](https://github.com/WCSim/WCSim/assets/19434001/72936f72-cf1a-49b3-bdf6-ec3581b0e8f6)
![hEmeanCosmics](https://github.com/WCSim/WCSim/assets/19434001/c8f39f48-a056-46b0-9200-d4e57fa08eac)

There was also a problem finding the cosmic muon energy from the mean energy histogram. The muon energy distribution didn't make sense but it looks like expected now:

![wcsim_HK_cosmics_histos_genE](https://github.com/WCSim/WCSim/assets/19434001/8792ddd1-0e86-4241-aa21-be4e0c2cd11e)

I'm assuming the input text file is correct. There are no muons below 100 GeV in the plot and that's compatible with the text file. I'm guessing that was made on purpose to study high energy cosmic muons. I don't know who created the input text file.

Lastly, all muons were generated passing exactly through the centre of the detector. As a consequence, there were no Cherenkov photons from muons starting in a gap along Z = 0. The photon start position looked like this after generating 300 cosmic muons:

![wcsim_HK_cosmics_run_start_rz](https://github.com/WCSim/WCSim/assets/19434001/e2584d2b-02a4-4bbe-8118-95785e094b96)

The code I added generates muons passing through a random point in the ID volume. The same distribution doesn't have the gap anymore (1000 muons):

![wcsim_HK_cosmics_merged_histos_start_rz](https://github.com/WCSim/WCSim/assets/19434001/8ae6416d-808f-4936-b8db-11a56cfc88ba)

tdealtry commented:
All looks great. Thanks @arturof
